swap/swap.go

```diff
@@ -179,6 +179,146 @@ func (s *Swap) Add(amount int64, peer *protocols.Peer) (err error) {
 	return nil
 }
 
+// handleMsg is for handling messages when receiving messages
+func (s *Swap) handleMsg(p *Peer) func(ctx context.Context, msg interface{}) error {
+	return func(ctx context.Context, msg interface{}) error {
+		switch msg := msg.(type) {
+		case *EmitChequeMsg:
+			go s.handleEmitChequeMsg(ctx, p, msg)
+		}
+		return nil
+	}
+}
+
+// handleEmitChequeMsg should be handled by the creditor when it receives
+// a cheque from a debitor
+func (s *Swap) handleEmitChequeMsg(ctx context.Context, p *Peer, msg *EmitChequeMsg) error {
+	cheque := msg.Cheque
+	log.Info("received cheque from peer", "peer", p.ID().String())
+	actualAmount, err := s.processAndVerifyCheque(cheque, p)
+	if err != nil {
+		return err
+	}
+
+	log.Debug("received cheque processed and verified", "peer", p.ID().String())
+
+	// reset balance by amount
+	// as this is done by the creditor, receiving the cheque, the amount should be negative,
+	// so that updateBalance will calculate balance + amount which result in reducing the peer's balance
+	err = s.resetBalance(p.ID(), 0-int64(cheque.Honey))
```

Comment:
This was not introduced in this PR but re-reading this we should probably lock s here (or inside `resetBalance`).

Reply:
Great catch man, I agree. Will post an update